File: python/triangle/triangle.py
def equilateral(sides: list) -> bool: 
    return not has_0_side(sides) and sides[0] == sides[1] == sides[2]


def isosceles(sides: list) -> bool: 
    return not has_0_side(sides) and (1 <= len(set(sides)) <= 2) and is_triangle(sides)


def scalene(sides: list) -> bool:  
    return not has_0_side(sides) and len(set(sides)) == 3 and is_triangle(sides)
    

def has_0_side(sides: list) -> bool:
    return 0 in sides


def is_triangle(sides: list) -> bool:
    return ((sides[0] + sides[1] >= sides[2]) 
            and (sides[0] + sides[2] >= sides[1])
            and (sides[1] + sides[2] >= sides[0]))

File: typescript/triangle/triangle.ts
export class Triangle {
  private sides: number[];

  constructor(...sides: number[]) {
    this.sides = sides;
  }

  get isEquilateral() {
    return this.sides.length === 3 
      && this.sides[0] !== 0
      && this.sides.every(s => s == this.sides[0]);
  }

  get isIsosceles() {
    if (this.sides.length !== 3 || this,this.sides.includes(0)) return false;
    if (this.isEquilateral) return true;

    if(!isTriangle(this.sides[0], this.sides[1], this.sides[2])) return false;

    return this.sides[0] === this.sides[1] 
      || this.sides[0] === this.sides[2]
      || this.sides[1] === this.sides[2];
  }

  get isScalene() {
    if (this.sides.length !== 3 || this,this.sides.includes(0)) return false;
    if (this.isEquilateral || this.isIsosceles) return false;
    return isTriangle(this.sides[0], this.sides[1], this.sides[2]);
  }
}

function isTriangle(a: number, b: number, c: number) :boolean {
  return ((a + b) >= c)
    && ((b + c) >= a)
    && ((a + c) >= b);
}



